### PR TITLE
record network stats if a DataDog client is passed in

### PR DESCRIPTION
**Pull request by craig-day**
Record some stats around network requests.

/cc @grosser

### CODE REVIEW

File: lib/i18n/backend/http/etag_http_client.rb [hunk 1]
```diff
@@ -6,24 +6,49 @@ class Http
...
         end
+
+        private
+
+        def record(event, options = {})
+          return unless @statsd_client
+
+          case event
+          when :success
+            @statsd_client.increment('i18n-backend-http.etag_client.success', tags: options[:tags])
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: extract `i18n-backend-http.etag_client.` into a constant ?

File: lib/i18n/backend/http/etag_http_client.rb [hunk 2]
```diff
@@ -6,24 +6,49 @@ class Http
...
+          when :failure
+            @statsd_client.increment('i18n-backend-http.etag_client.failure', tags: options[:tags])
+          when :timing
+            @statsd_client.histogram('i18n-backend-http.etag_client.request_time', options[:time], tags: options[:tags])
+          end
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: `else raise` ?

File: lib/i18n/backend/http/etag_http_client.rb [hunk 3]
```diff
@@ -6,24 +6,49 @@ class Http
       class I18n::Backend::Http::EtagHttpClient
         def initialize(options)
           @options = options
+          @statsd_client = options[:statsd_client]
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: leave it in `@options` and pull it out on demand ?

File: lib/i18n/backend/http/etag_http_client.rb [hunk 4]
```diff
@@ -6,24 +6,49 @@ class Http
...
           response = @client.get(path) do |request|
             request.headers.merge!(@options[:headers]) if @options[:headers]
             request.headers["If-None-Match"] = etag if etag
-            request.options[:timeout] = @options[:http_read_timeout]
-            request.options[:open_timeout] = @options[:http_open_timeout]
+            request.options[:timeout]        = @options[:http_read_timeout]
+            request.options[:open_timeout]   = @options[:http_open_timeout]
           end
 
+          record :timing, time: (Time.now - start).to_i, tags: {path: path}
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: report .to_f diff * 1000 to have milliseconds

File: lib/i18n/backend/http/etag_http_client.rb [hunk 5]
```diff
@@ -6,24 +6,49 @@ class Http
...
+
           case response.status
-          when 200 then [response.body, response['ETag']]
-          when 304 then nil
+          when 200
+            record :success, tags: {status_code: response.status, path: path}
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: path is a limited set and won't spam datadog, right ?

Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: ... maybe just send the .split(/).last ?

Comment (craig-day) [inline lib/i18n/backend/http/etag_http_client.rb]: I could make it nicer, but didn't want to get too zendesk specific. `path` is implemented by the consumer so hard to say, but for use it is limited.

**Comment by grosser** [comment]
should be fine ...

**Comment by craig-day** [comment]
@grosser any other thoughts?

File: lib/i18n/backend/http/etag_http_client.rb [hunk 6]
```diff
@@ -4,26 +4,52 @@ module I18n
...
           response = @client.get(path) do |request|
             request.headers.merge!(@options[:headers]) if @options[:headers]
             request.headers["If-None-Match"] = etag if etag
-            request.options[:timeout] = @options[:http_read_timeout]
-            request.options[:open_timeout] = @options[:http_open_timeout]
+            request.options[:timeout]        = @options[:http_read_timeout]
+            request.options[:open_timeout]   = @options[:http_open_timeout]
           end
 
+          record :timing, time: (Time.now - start).to_f * 1000, tags: {path: path}
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: tag this with status too so we can see if failures are slow / 304 are fast etc

File: lib/i18n/backend/http/etag_http_client.rb [hunk 7]
```diff
@@ -4,26 +4,52 @@ module I18n
...
+          return unless client = @options[:statsd_client]
+
+          case event
+          when :success
+            client.increment("#{STATS_NAMESPACE}.success", tags: options[:tags])
+          when :failure
+            client.increment("#{STATS_NAMESPACE}.failure", tags: options[:tags])
+          when :timing
+            client.histogram("#{STATS_NAMESPACE}.request_time", options[:time], tags: options[:tags])
+          end
```
Comment (grosser) [inline lib/i18n/backend/http/etag_http_client.rb]: `else raise` ?

Comment (craig-day) [inline lib/i18n/backend/http/etag_http_client.rb]: raise seems weird since it is a private methods... but I'll add an uncovered line if you want a raise

**Comment by grosser** [comment]
I think having a safety net instead of running into nils/noops when
something is misconfigured is better

**Pull request merged**